docs/api/parser.ts

```diff
@@ -0,0 +1,202 @@
+import Ast from 'ts-simple-ast'
+import {
+  SourceFile,
+  TypeAliasDeclaration,
+  JSDoc,
+  ClassDeclaration,
+  MethodDeclaration,
+  VariableDeclaration,
+  VariableStatement,
+  InterfaceDeclaration,
+  FunctionDeclaration
+} from 'ts-simple-ast'
+import { Module, Data, Constructor, Export, Method, Func, Typeclass, Instance } from './domain'
+import { Option, none, some, fromNullable } from '../../src/Option'
+import { catOptions } from '../../src/Array'
+import { parse, Tag, Annotation } from 'doctrine'
+
+// TODO: avoid comment duplication in overloaded function declarations
+
+export const parseJSDoc = (source: string) => parse(source, { unwrap: true })
+
+const getSourceFile = (name: string, source: string): SourceFile =>
+  new Ast().addSourceFileFromText(`${name}.ts`, source)
+
+const fromJSDocDescription = (description: string | null): Option<string> => {
+  const d = description ? description.trim() : ''
+  return d === '' ? none : some(d)
+}
+
+const getMethodSignature = (md: MethodDeclaration): string => {
+  const text = md.getText()
+  const start = md.getName().length
+  const end = text.indexOf('{')
+  return text.substring(start, end)
+}
+
+const getMethodDescription = (md: MethodDeclaration): Option<string> =>
+  fromNullable(md.getDocumentationComment()).filter(s => s.trim() !== '')
+
+const getMethod = (md: MethodDeclaration): Method => ({
+  type: 'method',
+  name: md.getName(),
+  signature: getMethodSignature(md),
+  description: getMethodDescription(md)
+})
+
+const getClassMethods = (cd: ClassDeclaration): Array<Method> => cd.getInstanceMethods().map(md => getMethod(md))
+
+const getAnnotation = (jsdocs: Array<JSDoc>): Annotation => parseJSDoc(jsdocs.map(doc => doc.getText()).join('\n'))
+
+const hasTag = (title: string) => (annotation: Annotation): boolean => annotation.tags.some(tag => tag.title === title)
+
+const isData = hasTag('data')
+
+const isConstructor = (tag: Tag) => tag.title === 'constructor'
+
+const isFunc = hasTag('function')
+
+const isInstance = hasTag('instance')
+
+const isAlias = hasTag('alias')
+
+const isTypeclass = hasTag('typeclass')
+
+const parseTypeAliasDeclarationData = (sf: SourceFile, tad: TypeAliasDeclaration): Option<Data> => {
+  const annotation = getAnnotation(tad.getDocumentationCommentNodes())
+  if (isData(annotation)) {
+    const dataName = tad.getName()
+    const signature = tad.getText().substring('export '.length)
+    const description = fromJSDocDescription(annotation.description)
+    const constructors: Array<Constructor> = annotation.tags.filter(tag => isConstructor(tag)).map(tag => {
+      const name = tag.name
+      if (typeof name === 'undefined') {
+        throw new Error(`Missing constructor name in ${dataName}`)
```

Comment:
Should pure fp library use `throw` in it's sources? Isn't it a poor example? ;)

Reply:
eh... I added "Quick and dirty" in the title specifically for that 😛 (and some poor string parsing..)

More seriously, if the result feels ok we can improve the code at a later stage

Reply:
Sounds good :)